src/Microsoft.DotNet.XHarness.WebAssembly/xunit-runner.cs

```diff
@@ -0,0 +1,354 @@
+        {
+            if (arg == null || arg.GetType() == argType)
+                return arg;
+            // Not clear what conversions should be done
+            // IEnumerable<T> -> T[]
+            if (argType.IsArray)
+            {
+                Type t = typeof(IEnumerable<>).MakeGenericType(argType.GetElementType());
+                if (t.IsAssignableFrom(arg.GetType()))
+                {
+                    MethodInfo m = typeof(Enumerable).GetMethod("ToArray").MakeGenericMethod(new Type[] { argType.GetElementType() });
+                    return m.Invoke(null, new object[] { arg });
+                }
+            }
+            try
+            {
+                if (arg != null && arg is IConvertible)
+                    arg = Convert.ChangeType(arg, argType);
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine("EX (ConvertArg): " + ex);
+            }
+
+            return arg;
+        }
+
+        public virtual bool OnMessage(IMessageSinkMessage msg)
+        {
+            if (msg is Xunit.Sdk.TestCaseDiscoveryMessage disc_msg)
+            {
+                _testCases.Add(disc_msg.TestCase);
+                return true;
+            }
+
+            return true;
+        }
+
+        public int Run()
+        {
+            Console.WriteLine("Running tests...");
+            Console.WriteLine(".");
+
+            foreach (ITestCase testCase in TestCases)
+            {
+                RunTestCase(testCase);
+            }
+
+            Console.WriteLine($"TESTS = { TestCases.Count() }, RUN = { nrun }, SKIP = { nfiltered }, FAIL = { nfail }");
+            return nfail == 0 ? 0 : 1;
+        }
+
+        private void RunTestCase(ITestCase testCase)
+        {
+            var tc = testCase as XunitTestCase;
+
+            if (!_project.Filters.Filter(tc))
+            {
+                nfiltered++;
+                return;
+            }
+
+            // var itrait_attrs = tc.TestMethod.Method.GetCustomAttributes(typeof(ITraitAttribute));
+            // // FIXME:
+            // if (itrait_attrs.Count () > 0) {
+            // 	Console.WriteLine ("SKIP (ITraitAttribute): " + tc.DisplayName);
+            // 	nfiltered ++;
+            // 	continue;
+            // }
+            /*
+            foreach (var attr in itrait_attrs) {
+                Console.WriteLine (attr);
+                foreach (var disc_attr in attr.GetCustomAttributes (typeof (TraitDiscovererAttribute))) {
+                    Console.WriteLine (disc_attr);
+
+                }
+            }
+            */
+
+            MethodInfo method = (tc.Method as ReflectionMethodInfo)?.MethodInfo;
+
+            if (method.ReflectedType.IsGenericTypeDefinition)
+            {
+                Console.WriteLine("FAIL (generic): " + tc.DisplayName);
+                nfail++;
+                return;
+            }
+
+            if (tc is XunitTheoryTestCase)
```

Comment:
This if is super large, lets create two helper methods, will be easier to follow the logic, if TheoryTestCase then Foo else Bar

Reply:
done